File: src/main/java/com/mxgraph/swing/util/mxAnimation.java
/**
 * Copyright (c) 2007, Gaudenz Alder
 */
package com.mxgraph.swing.util;

import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;

import javax.swing.Timer;

import com.mxgraph.util.mxEvent;
import com.mxgraph.util.mxEventObject;
import com.mxgraph.util.mxEventSource;

/**
 * Baseclass for all timer-based animations. Fires mxEvent.DONE when the
 * stopAnimation method is called. Implement updateAnimation for the
 * actual animation or listen to mxEvent.EXECUTE.
 */
public class mxAnimation extends mxEventSource
{
	/**
	 * Specifies the default delay for animations in ms. Default is 20.
	 */
	public static int DEFAULT_DELAY = 20;

	/**
	 * Default is DEFAULT_DELAY.
	 */
	protected int delay;

	/**
	 * Time instance that is used for timing the animation.
	 */
	protected Timer timer;

	/**
	 * Constructs a new animation instance with the given repaint delay.
	 */
	public mxAnimation()
	{
		this(DEFAULT_DELAY);
	}

	/**
	 * Constructs a new animation instance with the given repaint delay.
	 */
	public mxAnimation(int delay)
	{
		this.delay = delay;
	}

	/**
	 * Returns the delay for the animation.
	 */
	public int getDelay()
	{
		return delay;
	}

	/**
	 * Sets the delay for the animation.
	 */
	public void setDelay(int value)
	{
		delay = value;
	}
	
	/**
	 * Returns true if the animation is running.
	 */
	public boolean isRunning()
	{
		return timer != null;
	}

	/**
	 * Starts the animation by repeatedly invoking updateAnimation.
	 */
	public void startAnimation()
	{
		if (timer == null)
		{
			timer = new Timer(delay, new ActionListener()
			{

				public void actionPerformed(ActionEvent e)
				{
					updateAnimation();
				}

			});

			timer.start();
		}
	}

	/**
	 * Hook for subclassers to implement the animation. Invoke stopAnimation
	 * when finished, startAnimation to resume. This is called whenever the
	 * timer fires and fires an mxEvent.EXECUTE event with no properties.
	 */
	public void updateAnimation()
	{
		fireEvent(new mxEventObject(mxEvent.EXECUTE));
	}

	/**
	 * Stops the animation by deleting the timer and fires mxEvent.DONE.
	 */
	public void stopAnimation()
	{
		if (timer != null)
		{
			timer.stop();
			timer = null;
			fireEvent(new mxEventObject(mxEvent.DONE));
		}
	}

}

File: src/main/webapp/mxgraph/src/util/mxAnimation.js
/**
 * Copyright (c) 2006-2015, JGraph Ltd
 * Copyright (c) 2006-2015, Gaudenz Alder
 */
/**
 *
 * Class: mxAnimation
 * 
 * Implements a basic animation in JavaScript.
 * 
 * Constructor: mxAnimation
 * 
 * Constructs an animation.
 * 
 * Parameters:
 * 
 * graph - Reference to the enclosing <mxGraph>.
 */
function mxAnimation(delay)
{
	this.delay = (delay != null) ? delay : 20;
};

/**
 * Extends mxEventSource.
 */
mxAnimation.prototype = new mxEventSource();
mxAnimation.prototype.constructor = mxAnimation;

/**
 * Variable: delay
 * 
 * Specifies the delay between the animation steps. Defaul is 30ms.
 */
mxAnimation.prototype.delay = null;

/**
 * Variable: thread
 * 
 * Reference to the thread while the animation is running.
 */
mxAnimation.prototype.thread = null;

/**
 * Function: isRunning
 * 
 * Returns true if the animation is running.
 */
mxAnimation.prototype.isRunning = function()
{
	return this.thread != null;
};

/**
 * Function: startAnimation
 *
 * Starts the animation by repeatedly invoking updateAnimation.
 */
mxAnimation.prototype.startAnimation = function()
{
	if (this.thread == null)
	{
		this.thread = window.setInterval(mxUtils.bind(this, this.updateAnimation), this.delay);
	}
};

/**
 * Function: updateAnimation
 *
 * Hook for subclassers to implement the animation. Invoke stopAnimation
 * when finished, startAnimation to resume. This is called whenever the
 * timer fires and fires an mxEvent.EXECUTE event with no properties.
 */
mxAnimation.prototype.updateAnimation = function()
{
	this.fireEvent(new mxEventObject(mxEvent.EXECUTE));
};

/**
 * Function: stopAnimation
 *
 * Stops the animation by deleting the timer and fires an <mxEvent.DONE>.
 */
mxAnimation.prototype.stopAnimation = function()
{
	if (this.thread != null)
	{
		window.clearInterval(this.thread);
		this.thread = null;
		this.fireEvent(new mxEventObject(mxEvent.DONE));
	}
};
